### PR TITLE
Fixed autocomplete view positioning

### PR DESCRIPTION
**Pull request by jacekbogdanski**
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

View positioning was based on window pane size and scroll position. Now it depends on absolute rectangle of the editor frame.

### CODE REVIEW

**Comment by jacekbogdanski** [comment]
I refactored some missing semicolons etc from test file (looks like my linter wasn't working). There may be changes not related to this issue but i think it's good idea to just fix it now so we don't forgot about it in other PRs.

File: plugins/autocomplete/plugin.js [hunk 1]
```diff
@@ -718,36 +718,30 @@
...
+			var editor = this.editor,
+				viewPanelHeight = this.element.getSize( 'height' ),
+				windowFrame = editor.window.getFrame(),
+				editable = editor.editable(),
 				// Bounding rect where view should fit (visible editor viewport).
-				absoluteRect = editable.isInline() ? editable.getAbsoluteClientRect( this.editor ) : frame.getAbsoluteClientRect( this.editor ),
+				editorViewportRect = editable.isInline() ? editable.getAbsoluteClientRect( editor ) : windowFrame.getAbsoluteClientRect( editor ),
+				// How much space is there for the panel above and below the specified rect.
+				spaceAbove = rect.top - editorViewportRect.top,
+				spaceBelow = rect.bottom - editorViewportRect.bottom,
```
Comment (mlewand) [inline plugins/autocomplete/plugin.js]: Shouldn't the space below be computed by `editorViewportRect.bottom` - `rect.bottom`? Surprisingly I don't see any visible issues related to this value being computed incorrectly.

**Comment by jacekbogdanski** [comment]
About not addressed line here: https://github.com/ckeditor/ckeditor-dev/blob/3e782b9/tests/plugins/autocomplete/view.js#L226 (see #1732 (review))

I'm not sure what more could I change here? I already changed `config.absoluteRect` into `config.editorViewportRect` and renamed `elementStub` into `getClientRectStub`. Do you see any improvements which can I apply here?

**Comment by mlewand** [comment]
There's a bit of mess in commit history. We need to wait till #1727 lands in the major.

**Comment by jacekbogdanski** [comment]
I rebased it into the latest `t/1703` so it contains #1727 changes.

**Comment by jacekbogdanski** [comment]
I changed the manual test for `inline` editor - it has long content but I wanted to make sure that the editor is at the end of the browser viewport so it could be property tested.

File: tests/plugins/autocomplete/manual/viewposition.html [hunk 2]
```diff
@@ -0,0 +1,101 @@
+<style>
+	/* Swap margin with padding to prevent moving autocomplete panel outside editor. */
+	body {
+		margin-left: 0px;
```
[the remark this answers is not in the document]
Comment (jacekbogdanski) [reply tests/plugins/autocomplete/manual/viewposition.html]: That's true. I will try to adjust caret position relative to the document body. I think it's an edge case where a user adds margin into body element but it could be probably fixed without huge work.

**Comment by jacekbogdanski** [comment]
After some time fighting with margins on host body I gave up because I couldn't find an easy way to fix it.  I tried to remove preset body position from caret rect but it wasn't working well with different cases. I'm not sure if this is an important issue so I left margin hack from https://github.com/ckeditor/ckeditor-dev/pull/1732#discussion_r182790235 comment unchanged.

I don't see much use cases where someone would like to change body margins. I think this issue exists for `balloonpanel` also. WDYT @mlewand? Can we live with it or should I spend more time on fixing this issue?

**Comment by mlewand** [comment]
I have pushed a single commit with some changes that I have already been changing while reviewing.

**Comment by jacekbogdanski** [comment]
I moved view positioning tests into a separate file so it could be easily tested with different editors. Also, I simplified `setPosition` function so it doesn't contain multiple ifs. Instead, I'm recalculating view position for different cases so it's easier to read but adds additional performance impact (position is calculated more times). I think it's not worth to preoptimize this code and add complicated if-else conditions but would like to know your opinion @mlewand. 

My changes also closes #1913 and #1912.

**Comment by jacekbogdanski** [comment]
Rebased into the latest `t/1703`

**Comment by mlewand** [comment]
The only missing part there was ignoring tests on unsupported IE.
